### PR TITLE
dialects: (riscv) Add lowering from register-allocated rvscf to riscv labels

### PR DESCRIPTION
**Pull request by AntonLydike**
Whole pipeline runs: `python xdsl/transforms/rvscf_regalloc.py | xdsl-opt -p lower-riscv-scf-to-labels,lower-riscv-func -t riscv-asm`

This takes a small example, register allocates, lowers the scf to labels, and then prints assembly. Very rough around the edges.

When we elide unused moves and fold `li, add -> addi` we have one instruction less than gcc on `-O3`, whatever that's worth^^

Lots of improvement potential here as usual. 

@tobiasgrosser you may find this interesting.

### CODE REVIEW

File: xdsl/backend/riscv/riscv_scf_to_asm.py
Comment (superlopuh) [inline xdsl/backend/riscv/riscv_scf_to_asm.py:72]: Is there a way to use the rewriter for this? Maybe before detaching the block?

**Review by superlopuh — CHANGES_REQUESTED** [review]
Looks great overall, some small things to discuss. I would love @tobiasgrosser and @compor's reviews on this PR before merging

**Review by nazavode — APPROVED** [review]

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (kingiler) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:15]: Can use the output from `xdsl-opt tests/filecheck/backend/rvscf_lowering_labels.mlir --print-op-generic`

File: xdsl/backend/riscv/riscv_scf_to_asm.py
Comment (adutilleul) [reply xdsl/backend/riscv/riscv_scf_to_asm.py:14]: Inlining it requires a bit more code (duplication), I''m fine with either.

Comment (superlopuh) [reply xdsl/backend/riscv/riscv_scf_to_asm.py:14]: Let's leave it as-is

**Review by superlopuh — APPROVED** [review]

File: tests/filecheck/with-riscemu/rvscf_lowering_emu.mlir
Comment (kingiler) [inline tests/filecheck/with-riscemu/rvscf_lowering_emu.mlir:18]: I'm not sure about the use of `custom_assembly_instruction` here?

Comment (adutilleul) [reply tests/filecheck/with-riscemu/rvscf_lowering_emu.mlir:18]: As far as I know, this is the only way of getting the output from `riscemu` in the current filecheck infrastructure, see the other tests in the folder for reference.

Comment (kingiler) [reply tests/filecheck/with-riscemu/rvscf_lowering_emu.mlir:18]: I see. 93 is syscall for exit instead of printing.

**Review by kingiler — APPROVED** [review]

**Comment by compor** [comment]
> Looks great overall, some small things to discuss. I would love @tobiasgrosser and @compor's reviews on this PR before merging

will do tomorrow; no time today I'm afraid.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:7]: I am surprised that the input test case uses `riscv.mv` instructions? I feel these just are not needed and could be removed to reduce noise? Or do I miss sth here?

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:10]: Can we make this an integer loop? That would allow removing the fcvt instructions and would result in a more straightforward loop body. Also, given that floating-point is an extension over the base instruction set, maybe we should avoid using it in the base test cases for features that can work without floating point support

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:11]: I feel riscv_scf.for could really benefit from custom formatting, such that it is clear what the lb and upper bound are.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:33]: I guess the riscv.get_register() is how we move SSA-values around branches. That's an interesting approach that is different to what we did in ChocoPy. It mixes SSA-semantics within BBs and side-effectful semantics. I guess it also means that all instructions that write to register-allocated types are suddenly side-effectful as they will have write effects. Has this been documented somewhere? I feel getting this consistent might be difficult.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:72]: Nice that there is an example with nesting. I like it. I would like to see an example with frep, but maybe in a separate PR?

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [inline tests/filecheck/backend/rvscf_lowering_labels.mlir:50]: Is there a reason we are using --split-input-file? AFAIU this makes it harder to run individual test cases with `xdsl-opt`, no?

**Review by tobiasgrosser — COMMENTED** [review]
Hi @superlopuh, thank you for this PR. I think this level of PR is great to get things moving. I added some review comments. Please don't feel blocked by them, as I am a little bit out of this codebase so just want to add some ideas. Some of them might be better considered in subsequent PRs. Having a baseline test case of what we do today may already be useful.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:7]: This is an artefact of the register allocation approach. The idea is that the calling convention requires the funciton arguments to be in `a{k}` but we might want to reuse those registers inside the function. So the first thing we do is move them to a new register, which might end up being the same one, in which case the move will be eliminated during canonicalization.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:11]: Good idea, I filed an issue #1454

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:72]: Yep, let's do this in another PR, we still don't have all the operations defined that we would need for that test, I don't think

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:50]: It's a tradeoff, using `--split-input-file` lets us reuse the loaded instance of the xdsl module, which takes hundreds of ms to import, which adds up, but more importantly it grous the tests by semantical meaning, so I find it a bit easier to navigate. You're right that it has its drawbacks also.

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:10]: I'll add another test with an integer loop in another PR, I think it would be good to keep both.

**Comment by superlopuh** [comment]
Thank you everyone, merging

File: tests/filecheck/backend/rvscf_lowering_labels.mlir
Comment (tobiasgrosser) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:7]: I feel it would be useful to drop this leftover artefacts from register allocation in this test case. This will make the test easier to understand.

Comment (superlopuh) [reply tests/filecheck/backend/rvscf_lowering_labels.mlir:7]: https://github.com/xdslproject/xdsl/pull/1456

**Pull request merged**
